ring/lifecycler.go

```diff
@@ -586,22 +586,48 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 			return ringDesc, true, nil
 		}
 
+		level.Info(i.logger).Log("msg", "existing entry found in ring", "state", i.GetState(), "tokens", len(instanceDesc.Tokens), "ring", i.RingName)
+
 		// If the ingester failed to clean its ring entry up in can leave its state in LEAVING
 		// OR unregister_on_shutdown=false
 		// Move it into ACTIVE to ensure the ingester joins the ring.
-		if instanceDesc.State == LEAVING && len(instanceDesc.Tokens) == i.cfg.NumTokens {
-			instanceDesc.State = ACTIVE
-		}
+		if instanceDesc.State == LEAVING {
+			var tokens Tokens
+			isActive := true
+			if len(instanceDesc.Tokens) != i.cfg.NumTokens {
+				level.Debug(i.logger).Log("msg", "existing entry has different number of tokens", "existingTokens", len(instanceDesc.Tokens), "newTokens", i.cfg.NumTokens)
+				if len(tokensFromFile) > 0 {
+					level.Debug(i.logger).Log("msg", "adding tokens from file", "tokens", len(tokensFromFile))
+					isActive = len(tokensFromFile) >= i.cfg.NumTokens
```

Comment:
According to Owen the number of tokens may have gone down, wouldn't we leave the instance (~ingester) in LEAVING state?

If I understand correctly in case the instance wasn't in the ring, then this would lead to PENDING state (as that's where we start out from) and then transition to ACTIVE after a period of time.

Reply:
@krajorama You could find confirmation (via a test) that the proposed does indeed not lead to leaving state when the number of tokens has gone down?

> So I've looked at a test with decreasing number of tokens and the state is ACTIVE, however the number of tokens is not actually reduced, remains 256, not sure if this is ok or not - at least does not get stuck

You mean you wrote a test for the scenario?

Reply:
yes, and I see you've done the same. And I see you wait until the number goes down as well.